src/fides/api/task/execute_request_tasks.py

```diff
@@ -0,0 +1,403 @@
+    run_node_function: Task,
+    next_step: CurrentStep,
+    privacy_request_proceed: bool,
+) -> None:
+    """Queue downstream tasks of the current node **if** the downstream task has all its upstream tasks completed.
+
+    If we've reached the terminator task, restart the privacy request from the appropriate checkpoint.
+    """
+    pending_downstream: Query = request_task.get_pending_downstream_tasks(session)
+    for downstream_task in pending_downstream:
+        if downstream_task.upstream_tasks_complete(session):
+            logger_method(downstream_task)(
+                "Queuing {} task {} from {} {}: upstream nodes complete. Privacy Request: {}, Request Task {}.",
+                downstream_task.action_type.value,
+                downstream_task.collection_address,
+                request_task.action_type.value,
+                request_task.collection_address,
+                privacy_request.id,
+                downstream_task.id,
+            )
+            run_node_function.delay(
+                privacy_request_id=privacy_request.id,
+                privacy_request_task_id=downstream_task.id,
+                privacy_request_proceed=privacy_request_proceed,
+            )
+        else:
+            logger.debug(
+                "Cannot yet queue {} task {} from {}. Privacy Request: {}, Request Task {}. Waiting for other upstream nodes.",
+                downstream_task.action_type.value,
+                request_task.collection_address,
+                downstream_task.collection_address,
+                privacy_request.id,
+                downstream_task.id,
+            )
+
+    if (
+        request_task.request_task_address == TERMINATOR_ADDRESS
+        and request_task.status != ExecutionLogStatus.complete
+    ):
+        # Only queue privacy request from the next step if we haven't reached the terminator before.
+        # Multiple pathways could mark the same node as complete, so we may have already reached the
+        # terminator node through a quicker path.
+        from fides.api.service.privacy_request.request_runner_service import (
+            queue_privacy_request,
+        )
+
+        if (
+            privacy_request_proceed
+        ):  # For Testing, this could be set to False, so we could just
+            # run one of the graphs and not the entire privacy request
+            queue_privacy_request(
+                privacy_request_id=privacy_request.id,
+                from_step=next_step.value,
+            )
+        request_task.update_status(session, ExecutionLogStatus.complete)
+
+
+@celery_app.task(base=DatabaseTask, bind=True)
+def run_access_node(
+    self: DatabaseTask,
+    privacy_request_id: str,
+    privacy_request_task_id: str,
+    privacy_request_proceed: bool = True,
+) -> None:
+    """Run an individual task in the access graph for DSR 3.0 and queue downstream nodes
+    upon completion if applicable"""
+    with self.get_new_session() as session:
```

Comment:
Are these the database connections we're worried about? What happens if we exceed the `task_engine_pool_size` limit?

Reply:
I'm worried about both fides database connections and external database connections! I do have an API endpoint that lets you manually queue a "pending" task for cases where something has gotten stuck and am looking into more automated catch-all tasks that can re-queue tasks.

Thanks for highlighting this and I will continue to look at ways to recover if we max out database connection limits.